### ISSUE TITLE
Velero randomly binds PVCs to available PVs present in cluster

### ISSUE THREAD

**Issue by navilg** [vmware-tanzu/velero#7528]
**What steps did you take and what happened:**

I have PVs with reclaim policy as Retain. So I deleted the namespace which deleted the PVCs and other namespace-scoped resources.
After namespace and namespace-scoped resources are deleted, I updated the existing PVs to bring it in Available state by removing claimRef from it.

After this, I triggered velero restore using below command. Excluded the pods to make sure podvolumes are not restored and pods get created by statefulsets and deployment.

```
velero create restore <restorename> --from-backup=<backupname> --include-namespaces=<namespacename> --exclude-resources=pods
```

After restore is completed, We see that PVCs which are restored in cluster is bound to incorrect PVs (randomly). I checked in backup files in bucket and I could see backup PVCs json file have `.spec.volumeName` has correct PV name. But when restored, somehow it get bind to incorrect PV.

**What did you expect to happen:**

**The following information will help us better understand what's going on**:

Since backup json file contain `.spec.volumeName` and the PV is already present in cluster, It should have bound to correct PV mentioned in `.spec.volumeName`.

_If you are using velero v1.7.0+:_  
Please use `velero debug  --backup <backupname> --restore <restorename>` to generate the support bundle, and attach to this issue, more options please refer to `velero debug --help` 

_If you are using earlier versions:_  
Please provide the output of the following commands (Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)
- `kubectl logs deployment/velero -n velero`
- `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`
- `velero backup logs <backupname>`
- `velero restore describe <restorename>` or `kubectl get restore/<restorename> -n velero -o yaml`
- `velero restore logs <restorename>`

**Anything else you would like to add:**

**Environment:**

- Velero version (use `velero version`): 1.12.1 as well as 1.10.2
- Velero features (use `velero client config get features`): 
- Kubernetes version (use `kubectl version`): 1.28
- Kubernetes installer & version: GKE
- Cloud provider or hardware configuration: GCP
- OS (e.g. from `/etc/os-release`): Ubuntu

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

**Comment by sseago** [comment]
In most cases, VolumeName is removed from the PVC because the expectation is that a new PV will be provisioned (or chosen from available PVs if you're not using dynamic provisioning). The new volume should get its contents from either the snapshot or the filesystem backup, depending on how you backed up the volumes in the first place.

Looking at the code just now, it seems that we should only be removing VolumeName if we have a snapshot for the volume, we have a PodVolumeBackup (fs backup) for the volume, or the retain policy is delete.

Since your policy is retain, you should determine whether the associated PVCs have a snapshot or filesystem backup created for them.

**Comment by reasonerjt** [comment]
So you mean in the backup you backup the data in the PV, but intentionally want the PVC to be rebound to the existing PV?

IMO this is a relatively corner case, let me put this in icebox.

**Comment by sseago** [comment]
@blackpiglet yes, that should work. As long as we're not creating snapshots or fs backups (and PV policy is not `Delete`), then Velero should restore PVC without changing PV reference. We've used this for migration use cases with NFS volumes -- PVC and PV are backed up without snapshots or fs-backup, then restored to a new cluster where PVC and PV definitions are restored intact, pointing to the same external NFS share as before backup.

**Comment by navilg** [comment]
I tested below two scenarios:

**Scenario 1**:  Backup with --default-volumes-to-restic=false and --snapshot-volumes=false flags in velero backup command to exclude pod volumes in backup. PVs here have reclaim policy as Retain.
1. I deleted the application namespace. All pods, PVCs and other namespace scoped resources got deleted.
2. PVs went to Released state.
3. I made all the PV to available state by removing claimRef from PVs.
4. Restored from the backup.
5. After backups are restored, I see PVCs are bound correctly and application works fine.

**Scenario 2**: Backup without the flags added in scenario 1. This backs up all resources along with podvolumes.
1. Performed same restore step as in scenario 1 except in velero restore command, I used --exclude-resources=pods so that podvolumes are not restored
2. After backups are restored, PVCs gets bind to incorrect PVs.
3. Apps does not work fine.

When I see in GCS bucket inside bucket folder in tar.gz file, I can see volumeName field is present inside persistentvolumeclaim for both the backups. So is it like, It removes this reference during restore ?

**Comment by sseago** [comment]
@navilg What is the server default for default-volumes-to-restic? If you don't specify false or true for the backup, then the default is used -- check that field on thebackup CR to see what it used. It's possible it was 'true'. You definitely want it set to false for your use case. If Velero has a podvolumebackup (generated when default-volumes-to-restic -- or rather,  default-volumes-to-fs-backup, since the restic-specific one is deprecated -- is true), then velero resets volumename on restore so it can restore the volume. If your backup has a PodVolumeBackup, the only way to skip this completely would be to exclude the PVC itself from restore. But since it sounds like you don't want to back up volume content with fs-backup here, the change needed would be on the backup side.

Also, note that if you aren't performing fs-backup (i.e. that flag is false), then unless snapshot-volumes is also false, Velero will attempt to snapshot the volume via CSI or one of the volume snapshotter plugins if the volume is compatible. So if your volumes support snapshots (and you don't want to restore *new* volumes to restore from snapshot), you also need --snapshot-volumes to be false.

**Comment by navilg** [comment]
Thanks @sseago 

Server default for default-volumes-to-restic (default-volume-tofs) is set to true.
So here is my use case:

My scheduled backup runs on its schedule which includes pod volumes backup.
One fine bad Monday, I need to restore the application from last scheduled backup, but I don't want to restore PV's data because mess is only on resources (deployment and other resources) and my PV data are still good.
is there a way I can do that, means, Can I just restore resources excluding pod volumes from the backup which has pod volume as well as resources in it ?

**Comment by sseago** [comment]
@navilg Ok, so yes. If you want to restore only some resource types, then you can use the includedResources field. On the Backup side, includes/excludes tell Velero what to back up from the cluster. On the restore side, includes/excludes tell Velero what to restore from the backup.

So if you only want to restore Deployments (and other specified resources), then just include these. If you want everything but volume data, then you can exclude PVs, PVCs, and Pods using the excludedResources.
